File: examples/java/src/main/java/com/docflow/EntityExtraction.java
package com.docflow;

import com.google.gson.*;
import okhttp3.*;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * TextIn 智能文档抽取 API 示例
 *
 * <p>演示两种抽取模式：
 * <ol>
 *   <li>Prompt 模式：提供自然语言 prompt，由大模型理解并自由抽取</li>
 *   <li>字段模式：提供结构化的字段列表，进行精确可控抽取</li>
 * </ol>
 *
 * <p>依赖：OkHttp 4.x、Gson（见 pom.xml）
 *
 * <p>运行方式：
 * <pre>
 *   cd examples/java
 *   mvn clean package -q
 *   java -cp target/docflow-examples-1.0.0.jar com.docflow.EntityExtraction &lt;文件路径&gt;
 * </pre>
 */
public class EntityExtraction {

    // ============================================================
    // 配置项 — 请替换为您的实际值
    // ============================================================
    private static final String APP_ID      = "your-app-id";       // x-ti-app-id
    private static final String SECRET_CODE = "your-secret-code";  // x-ti-secret-code

    private static final String API_URL = "https://api.textin.com/ai/service/v2/entity_extraction";

    // ============================================================
    // 全局工具对象
    // ============================================================
    private static final OkHttpClient HTTP = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(120, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();

    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
    private static final MediaType JSON_TYPE = MediaType.get("application/json; charset=utf-8");

    // ============================================================
    // 工具辅助函数
    // ============================================================

    /** 构造鉴权请求头。 */
    private static Headers authHeaders() {
        return new Headers.Builder()
                .add("x-ti-app-id", APP_ID)
                .add("x-ti-secret-code", SECRET_CODE)
                .build();
    }

    /** 读取本地文件并转为 base64 字符串。 */
    private static String encodeFile(String filePath) throws IOException {
        byte[] bytes = Files.readAllBytes(new File(filePath).toPath());
        return Base64.getEncoder().encodeToString(bytes);
    }

    /**
     * 调用智能文档抽取 API。
     *
     * @param filePath    待抽取的本地文件路径
     * @param body        请求体（不含 file 字段）
     * @param urlParams   URL 参数，例如 {"parse_mode": "scan"}
     * @return 解析后的 JSON 响应
     * @throws IOException          网络或文件读取异常
     * @throws RuntimeException     API 返回非 200 错误码时抛出
     */
    private static JsonObject callApi(String filePath, JsonObject body, Map<String, String> urlParams)
            throws IOException {
        body.addProperty("file", encodeFile(filePath));

        HttpUrl.Builder urlBuilder = Objects.requireNonNull(HttpUrl.parse(API_URL)).newBuilder();
        if (urlParams != null) {
            urlParams.forEach(urlBuilder::addQueryParameter);
        }

        Request req = new Request.Builder()
                .url(urlBuilder.build())
                .headers(authHeaders())
                .post(RequestBody.create(GSON.toJson(body), JSON_TYPE))
                .build();

        try (Response resp = HTTP.newCall(req).execute()) {
            String respBody = resp.body().string();
            JsonObject result = JsonParser.parseString(respBody).getAsJsonObject();
            if (result.get("code").getAsInt() != 200) {
                throw new RuntimeException("API 错误 " + result.get("code") + "："
                        + result.get("message").getAsString());
            }
            return result;
        }
    }

    // ============================================================
    // 示例 1：Prompt 模式
    // ============================================================

    public static void demoPromptMode(String filePath) throws IOException {
        System.out.println("=".repeat(60));
        System.out.println("示例 1：Prompt 模式");
        System.out.println("=".repeat(60));

        JsonObject body = new JsonObject();
        body.addProperty("prompt", "请抽取文件中的所有关键字段，以 JSON 格式返回");

        Map<String, String> params = new HashMap<>();
        params.put("parse_mode", "scan");

        JsonObject result = callApi(filePath, body, params);
        JsonObject res = result.getAsJsonObject("result");

        // llm_json：大模型直接返回的键值对，方便直接使用
        JsonElement llmJson = res.get("llm_json");
        System.out.println("\n抽取结果 (llm_json)：");
        System.out.println(GSON.toJson(llmJson));

        // usage：token 消耗统计
        JsonObject usage = res.getAsJsonObject("usage");
        if (usage != null) {
            System.out.printf("%nToken 消耗：输入 %d，输出 %d，合计 %d%n",
                    usage.get("prompt_tokens").getAsInt(),
                    usage.get("completion_tokens").getAsInt(),
                    usage.get("total_tokens").getAsInt());
        }
    }

    // ============================================================
    // 示例 2：字段模式
    // ============================================================

    public static void demoFieldMode(String filePath) throws IOException {
        System.out.println("=".repeat(60));
        System.out.println("示例 2：字段模式");
        System.out.println("=".repeat(60));

        JsonObject body = new JsonObject();

        // fields：要抽取的单值字段列表
        JsonArray fields = new JsonArray();
        for (String name : new String[]{"发票号码", "开票日期", "购买方名称", "销售方名称", "合计金额"}) {
            JsonObject f = new JsonObject();
            f.addProperty("name", name);
            fields.add(f);
        }
        body.add("fields", fields);

        // table_fields：要抽取的表格字段（可选）
        JsonArray tableFields = new JsonArray();
        JsonObject table = new JsonObject();
        table.addProperty("title", "货物明细");
        table.addProperty("description", "发票中的货物或服务明细表格");
        JsonArray tableCols = new JsonArray();
        for (String col : new String[]{"项目名称", "数量", "单价", "金额"}) {
            JsonObject c = new JsonObject();
            c.addProperty("name", col);
            tableCols.add(c);
        }
        table.add("fields", tableCols);
        tableFields.add(table);
        body.add("table_fields", tableFields);

        Map<String, String> params = new HashMap<>();
        params.put("parse_mode", "scan");

        JsonObject result = callApi(filePath, body, params);
        JsonObject res = result.getAsJsonObject("result");

        // details：字段抽取结果
        JsonObject details = res.getAsJsonObject("details");
        System.out.println("\n抽取结果 (details)：");
        if (details != null) {
            for (Map.Entry<String, JsonElement> entry : details.entrySet()) {
                if ("row".equals(entry.getKey())) {
                    JsonArray rows = entry.getValue().getAsJsonArray();
                    System.out.println("  表格（货物明细）：共 " + rows.size() + " 行");
                    for (int i = 0; i < rows.size(); i++) {
                        System.out.println("    第 " + (i + 1) + " 行：" + rows.get(i));
                    }
                } else if (entry.getValue().isJsonObject()) {
                    String val = entry.getValue().getAsJsonObject().get("value").getAsString();
                    System.out.println("  " + entry.getKey() + "：" + val);
                }
            }
        }

        System.out.println("\n处理页数：" + res.get("page_count").getAsInt());
        System.out.println("推理时间：" + result.get("duration").getAsInt() + " ms");
    }

    // ============================================================
    // 入口
    // ============================================================

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("用法：java EntityExtraction <文件路径>");
            System.out.println("示例：java EntityExtraction invoice.pdf");
            System.exit(1);
        }

        String filePath = args[0];
        if (!new File(filePath).exists()) {
            System.out.println("文件不存在：" + filePath);
            System.exit(1);
        }

        demoPromptMode(filePath);
        System.out.println();
        demoFieldMode(filePath);
    }
}

File: examples/python/entity_extraction.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
TextIn 智能文档抽取 API 示例

演示两种抽取模式：
  1. Prompt 模式：提供自然语言 prompt，由大模型理解并自由抽取
  2. 字段模式：提供结构化的字段列表，进行精确可控抽取

支持格式：png, jpg, jpeg, pdf, bmp, tiff, webp, doc, docx, html,
          mhtml, xls, xlsx, csv, ppt, pptx, txt, ofd

依赖：
  pip install requests

使用前请先填写下方配置项，然后执行：
  python entity_extraction.py <文件路径>
"""

import base64
import json
import os
import sys

import requests

# ============================================================
# 配置项 — 请替换为您的实际值
# ============================================================
APP_ID      = "your-app-id"       # TextIn 控制台中的 x-ti-app-id
SECRET_CODE = "your-secret-code"  # TextIn 控制台中的 x-ti-secret-code

API_URL = "https://api.textin.com/ai/service/v2/entity_extraction"


# ============================================================
# 工具辅助函数
# ============================================================

def _headers() -> dict:
    """构造鉴权请求头。"""
    return {
        "x-ti-app-id":      APP_ID,
        "x-ti-secret-code": SECRET_CODE,
        "Content-Type":     "application/json",
    }


def _encode_file(file_path: str) -> str:
    """读取本地文件并转为 base64 字符串。"""
    with open(file_path, "rb") as f:
        return base64.b64encode(f.read()).decode("utf-8")


def extract(file_path: str, request_body: dict, url_params: dict = None) -> dict:
    """
    调用智能文档抽取 API。

    :param file_path:     待抽取的本地文件路径
    :param request_body:  请求体字段（不含 file 字段）
    :param url_params:    URL 参数（可选），例如 {"parse_mode": "scan"}
    :return:              解析后的 JSON 响应
    :raises RuntimeError: API 返回非 200 错误码时抛出
    """
    body = {**request_body, "file": _encode_file(file_path)}
    resp = requests.post(
        API_URL,
        params=url_params or {},
        headers=_headers(),
        json=body,
        timeout=120,
    )
    resp.raise_for_status()
    result = resp.json()
    if result.get("code") != 200:
        raise RuntimeError(f"API 错误 {result.get('code')}：{result.get('message')}")
    return result


# ============================================================
# 示例 1：Prompt 模式
# ============================================================

def demo_prompt_mode(file_path: str):
    """以 Prompt 模式对文档进行智能抽取。"""
    print("=" * 60)
    print("示例 1：Prompt 模式")
    print("=" * 60)

    result = extract(
        file_path,
        request_body={
            "prompt": "请抽取文件中的所有关键字段，以 JSON 格式返回",
        },
        url_params={"parse_mode": "scan"},
    )

    res = result.get("result", {})

    # llm_json：大模型直接返回的键值对，方便直接使用
    llm_json = res.get("llm_json", {})
    print("\n抽取结果 (llm_json)：")
    print(json.dumps(llm_json, ensure_ascii=False, indent=2))

    # usage：token 消耗统计
    usage = res.get("usage", {})
    if usage:
        print(f"\nToken 消耗：输入 {usage.get('prompt_tokens', 0)}，"
              f"输出 {usage.get('completion_tokens', 0)}，"
              f"合计 {usage.get('total_tokens', 0)}")


# ============================================================
# 示例 2：字段模式
# ============================================================

def demo_field_mode(file_path: str):
    """以自定义字段模式对文档进行精确抽取。"""
    print("=" * 60)
    print("示例 2：字段模式")
    print("=" * 60)

    result = extract(
        file_path,
        request_body={
            # fields：要抽取的单值字段列表
            "fields": [
                {"name": "发票号码"},
                {"name": "开票日期"},
                {"name": "购买方名称"},
                {"name": "销售方名称"},
                {"name": "合计金额"},
            ],
            # table_fields：要抽取的表格字段（可选）
            "table_fields": [
                {
                    "title": "货物明细",
                    "description": "发票中的货物或服务明细表格",
                    "fields": [
                        {"name": "项目名称"},
                        {"name": "数量"},
                        {"name": "单价"},
                        {"name": "金额"},
                    ],
                }
            ],
        },
        url_params={"parse_mode": "scan"},
    )

    res = result.get("result", {})

    # details：字段抽取结果
    details = res.get("details", {})
    print("\n抽取结果 (details)：")
    for key, val in details.items():
        if key == "row":
            rows = val if isinstance(val, list) else []
            print(f"  表格（货物明细）：共 {len(rows)} 行")
            for i, row in enumerate(rows, 1):
                print(f"    第 {i} 行：{json.dumps(row, ensure_ascii=False)}")
        elif isinstance(val, dict):
            print(f"  {key}：{val.get('value', '')}")

    print(f"\n处理页数：{res.get('page_count', '-')}")
    print(f"推理时间：{result.get('duration', '-')} ms")


# ============================================================
# 入口
# ============================================================

def main():
    if len(sys.argv) < 2:
        print("用法：python entity_extraction.py <文件路径>")
        print("示例：python entity_extraction.py invoice.pdf")
        sys.exit(1)

    file_path = sys.argv[1]
    if not os.path.exists(file_path):
        print(f"文件不存在：{file_path}")
        sys.exit(1)

    demo_prompt_mode(file_path)
    print()
    demo_field_mode(file_path)


if __name__ == "__main__":
    main()
